File: Infectious Disease Case Detection and Modeling Framework/scripts/r/ml_models.R
# Machine Learning Analysis Module
# R script for machine learning analysis

library(caret)
library(randomForest)
library(xgboost)
library(glmnet)
library(Metrics)
library(dplyr)
source("data_loader.R")

# Function to prepare ML data
# Prepare data for machine learning
# Parameters: df_clean - cleaned dataset, target - target variable name
prepare_ml_data <- function(df_clean, target = "Incidence_Rate") {
  # Select features
  feature_cols <- c("CI_Lower", "CI_Upper", "Annual_Count", "Trend_5yr", "FIPS")
  feature_cols <- feature_cols[feature_cols %in% colnames(df_clean)]
  
  # Handle categorical variables
  if ("Trend" %in% colnames(df_clean)) {
    df_clean$Trend_encoded <- as.numeric(as.factor(df_clean$Trend))
    feature_cols <- c(feature_cols, "Trend_encoded")
  }
  
  # Create feature matrix and target vector
  X <- df_clean[, feature_cols, drop = FALSE]
  y <- df_clean[[target]]
  
  # Remove rows with missing values
  complete_cases <- complete.cases(X, y)
  X <- X[complete_cases, , drop = FALSE]
  y <- y[complete_cases]
  
  # Split data
  set.seed(42)
  train_index <- createDataPartition(y, p = 0.8, list = FALSE)
  X_train <- X[train_index, , drop = FALSE]
  X_test <- X[-train_index, , drop = FALSE]
  y_train <- y[train_index]
  y_test <- y[-train_index]
  
  # Scale features
  preProc <- preProcess(X_train, method = c("center", "scale"))
  X_train_scaled <- predict(preProc, X_train)
  X_test_scaled <- predict(preProc, X_test)
  
  return(list(
    X_train = X_train_scaled,
    X_test = X_test_scaled,
    y_train = y_train,
    y_test = y_test,
    preProc = preProc,
    feature_cols = feature_cols
  ))
}

# Function to train and evaluate models
# Train and evaluate multiple ML models
# Parameters: X_train/X_test - training/test features, y_train/y_test - training/test targets
train_and_evaluate_models <- function(X_train, X_test, y_train, y_test) {
  models <- list()
  results <- data.frame()
  
  cat("=== MACHINE LEARNING MODEL EVALUATION ===\n\n")
  
  # Linear Regression
  cat("Training Linear Regression...\n")
  lm_model <- lm(y_train ~ ., data = data.frame(X_train, y_train = y_train))
  lm_pred_train <- predict(lm_model, newdata = data.frame(X_train))
  lm_pred_test <- predict(lm_model, newdata = data.frame(X_test))
  
  results <- rbind(results, data.frame(
    Model = "Linear Regression",
    Train_RMSE = rmse(y_train, lm_pred_train),
    Test_RMSE = rmse(y_test, lm_pred_test),
    Train_MAE = mae(y_train, lm_pred_train),
    Test_MAE = mae(y_test, lm_pred_test),
    Train_R2 = R2(lm_pred_train, y_train),
    Test_R2 = R2(lm_pred_test, y_test)
  ))
  models[["Linear Regression"]] <- lm_model
  
  # Ridge Regression
  cat("Training Ridge Regression...\n")
  ridge_model <- train(
    x = X_train, y = y_train,
    method = "ridge",
    trControl = trainControl(method = "cv", number = 5),
    tuneGrid = expand.grid(lambda = seq(0.001, 1, length = 10))
  )
  ridge_pred_train <- predict(ridge_model, newdata = X_train)
  ridge_pred_test <- predict(ridge_model, newdata = X_test)
  
  results <- rbind(results, data.frame(
    Model = "Ridge Regression",
    Train_RMSE = rmse(y_train, ridge_pred_train),
    Test_RMSE = rmse(y_test, ridge_pred_test),
    Train_MAE = mae(y_train, ridge_pred_train),
    Test_MAE = mae(y_test, ridge_pred_test),
    Train_R2 = R2(ridge_pred_train, y_train),
    Test_R2 = R2(ridge_pred_test, y_test)
  ))
  models[["Ridge Regression"]] <- ridge_model
  
  # Lasso Regression
  cat("Training Lasso Regression...\n")
  lasso_model <- train(
    x = X_train, y = y_train,
    method = "lasso",
    trControl = trainControl(method = "cv", number = 5),
    tuneGrid = expand.grid(fraction = seq(0.1, 1, length = 10))
  )
  lasso_pred_train <- predict(lasso_model, newdata = X_train)
  lasso_pred_test <- predict(lasso_model, newdata = X_test)
  
  results <- rbind(results, data.frame(
    Model = "Lasso Regression",
    Train_RMSE = rmse(y_train, lasso_pred_train),
    Test_RMSE = rmse(y_test, lasso_pred_test),
    Train_MAE = mae(y_train, lasso_pred_train),
    Test_MAE = mae(y_test, lasso_pred_test),
    Train_R2 = R2(lasso_pred_train, y_train),
    Test_R2 = R2(lasso_pred_test, y_test)
  ))
  models[["Lasso Regression"]] <- lasso_model
  
  # Random Forest
  cat("Training Random Forest...\n")
  rf_model <- randomForest(
    x = X_train, y = y_train,
    ntree = 100, mtry = sqrt(ncol(X_train)),
    importance = TRUE
  )
  rf_pred_train <- predict(rf_model, newdata = X_train)
  rf_pred_test <- predict(rf_model, newdata = X_test)
  
  results <- rbind(results, data.frame(
    Model = "Random Forest",
    Train_RMSE = rmse(y_train, rf_pred_train),
    Test_RMSE = rmse(y_test, rf_pred_test),
    Train_MAE = mae(y_train, rf_pred_train),
    Test_MAE = mae(y_test, rf_pred_test),
    Train_R2 = R2(rf_pred_train, y_train),
    Test_R2 = R2(rf_pred_test, y_test)
  ))
  models[["Random Forest"]] <- rf_model
  
  # Gradient Boosting
  cat("Training Gradient Boosting...\n")
  gbm_model <- train(
    x = X_train, y = y_train,
    method = "gbm",
    trControl = trainControl(method = "cv", number = 5),
    verbose = FALSE
  )
  gbm_pred_train <- predict(gbm_model, newdata = X_train)
  gbm_pred_test <- predict(gbm_model, newdata = X_test)
  
  results <- rbind(results, data.frame(
    Model = "Gradient Boosting",
    Train_RMSE = rmse(y_train, gbm_pred_train),
    Test_RMSE = rmse(y_test, gbm_pred_test),
    Train_MAE = mae(y_train, gbm_pred_train),
    Test_MAE = mae(y_test, gbm_pred_test),
    Train_R2 = R2(gbm_pred_train, y_train),
    Test_R2 = R2(gbm_pred_test, y_test)
  ))
  models[["Gradient Boosting"]] <- gbm_model
  
  # Print results
  print(results)
  
  return(list(
    models = models,
    results = results
  ))
}

# Function to plot model comparison
# Plot model comparison
plot_model_comparison <- function(results) {
  library(ggplot2)
  library(gridExtra)
  
  p1 <- ggplot(results, aes(x = reorder(Model, Test_R2), y = Test_R2)) +
    geom_bar(stat = "identity", fill = "steelblue", alpha = 0.7) +
    coord_flip() +
    labs(title = "Model Comparison: R² Score", x = "Model", y = "R² Score") +
    theme_minimal()
  
  p2 <- ggplot(results, aes(x = reorder(Model, -Test_RMSE), y = Test_RMSE)) +
    geom_bar(stat = "identity", fill = "steelblue", alpha = 0.7) +
    coord_flip() +
    labs(title = "Model Comparison: RMSE", x = "Model", y = "RMSE") +
    theme_minimal()
  
  png("ml_model_comparison.png", width = 1500, height = 600, res = 300)
  grid.arrange(p1, p2, ncol = 2)
  dev.off()
}

# Function for feature importance analysis
# Analyze feature importance
feature_importance_analysis <- function(model, feature_names) {
  if (class(model)[1] == "randomForest") {
    importance <- importance(model)
    importance_df <- data.frame(
      Feature = rownames(importance),
      Importance = importance[, "%IncMSE"]
    )
  } else if (class(model)[1] == "train" && model$method == "gbm") {
    importance <- varImp(model)
    importance_df <- data.frame(
      Feature = rownames(importance$importance),
      Importance = importance$importance$Overall
    )
  } else {
    cat("Model does not support feature importance\n")
    return(NULL)
  }
  
  importance_df <- importance_df[order(-importance_df$Importance), ]
  
  # Plot
  png("ml_feature_importance.png", width = 1000, height = 600, res = 300)
  ggplot(importance_df, aes(x = reorder(Feature, Importance), y = Importance)) +
    geom_bar(stat = "identity", fill = "steelblue", alpha = 0.7) +
    coord_flip() +
    labs(title = "Feature Importance", x = "Feature", y = "Importance") +
    theme_minimal()
  dev.off()
  
  return(importance_df)
}

# Main execution
if (!interactive()) {
  # Load and prepare data
  df <- load_data()
  df_clean <- clean_data(df)
  
  ml_data <- prepare_ml_data(df_clean)
  
  cat("Training set size:", nrow(ml_data$X_train), "\n")
  cat("Test set size:", nrow(ml_data$X_test), "\n")
  cat("Features:", ml_data$feature_cols, "\n\n")
  
  # Train and evaluate models
  results_list <- train_and_evaluate_models(
    ml_data$X_train, ml_data$X_test,
    ml_data$y_train, ml_data$y_test
  )
  
  # Plot comparison
  plot_model_comparison(results_list$results)
  
  # Feature importance for best model
  best_model_name <- results_list$results$Model[which.max(results_list$results$Test_R2)]
  best_model <- results_list$models[[best_model_name]]
  cat("\nBest Model:", best_model_name, "\n")
  importance_df <- feature_importance_analysis(best_model, ml_data$feature_cols)
  if (!is.null(importance_df)) {
    cat("\nFeature Importance:\n")
    print(importance_df)
  }
}

File: Infectious Disease Case Detection and Modeling Framework/scripts/python/ml_models.py
"""
Machine Learning Analysis Module
"""
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import xgboost as xgb
import lightgbm as lgb
import warnings
warnings.filterwarnings('ignore')

import sys
import os
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
from data_loader import load_data, clean_data

def prepare_ml_data(df_clean, target='Incidence_Rate'):
    """
    Prepare data for machine learning
    
    Parameters:
    -----------
    df_clean : DataFrame
        Cleaned dataset
    target : str
        Target variable
    """
    # Select features
    feature_cols = ['CI_Lower', 'CI_Upper', 'Annual_Count', 'Trend_5yr', 'FIPS']
    feature_cols = [col for col in feature_cols if col in df_clean.columns]
    
    # Handle categorical variables
    if 'Trend' in df_clean.columns:
        le = LabelEncoder()
        df_clean['Trend_encoded'] = le.fit_transform(df_clean['Trend'].astype(str))
        feature_cols.append('Trend_encoded')
    
    # Create feature matrix and target vector
    X = df_clean[feature_cols].dropna()
    y = df_clean.loc[X.index, target]
    
    # Remove rows with missing target
    mask = ~y.isna()
    X = X[mask]
    y = y[mask]
    
    # Scale features
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X)
    
    # Split data
    X_train, X_test, y_train, y_test = train_test_split(
        X_scaled, y, test_size=0.2, random_state=42
    )
    
    return X_train, X_test, y_train, y_test, scaler, feature_cols

def train_and_evaluate_models(X_train, X_test, y_train, y_test):
    """
    Train and evaluate multiple ML models
    
    Parameters:
    -----------
    X_train : array
        Training features
    X_test : array
        Test features
    y_train : array
        Training target
    y_test : array
        Test target
    """
    models = {
        'Linear Regression': LinearRegression(),
        'Ridge Regression': Ridge(alpha=1.0),
        'Lasso Regression': Lasso(alpha=1.0),
        'Elastic Net': ElasticNet(alpha=1.0, l1_ratio=0.5),
        'Random Forest': RandomForestRegressor(n_estimators=100, random_state=42),
        'Gradient Boosting': GradientBoostingRegressor(n_estimators=100, random_state=42),
        'XGBoost': xgb.XGBRegressor(random_state=42),
        'LightGBM': lgb.LGBMRegressor(random_state=42, verbosity=-1)
    }
    
    results = {}
    
    print("=== MACHINE LEARNING MODEL EVALUATION ===\n")
    
    for name, model in models.items():
        print(f"Training {name}...")
        
        # Train model
        model.fit(X_train, y_train)
        
        # Predictions
        y_pred_train = model.predict(X_train)
        y_pred_test = model.predict(X_test)
        
        # Metrics
        train_rmse = np.sqrt(mean_squared_error(y_train, y_pred_train))
        test_rmse = np.sqrt(mean_squared_error(y_test, y_pred_test))
        train_mae = mean_absolute_error(y_train, y_pred_train)
        test_mae = mean_absolute_error(y_test, y_pred_test)
        train_r2 = r2_score(y_train, y_pred_train)
        test_r2 = r2_score(y_test, y_pred_test)
        
        # Cross-validation
        cv_scores = cross_val_score(model, X_train, y_train, cv=5, scoring='r2')
        
        results[name] = {
            'model': model,
            'train_rmse': train_rmse,
            'test_rmse': test_rmse,
            'train_mae': train_mae,
            'test_mae': test_mae,
            'train_r2': train_r2,
            'test_r2': test_r2,
            'cv_mean': cv_scores.mean(),
            'cv_std': cv_scores.std()
        }
        
        print(f"  Train RMSE: {train_rmse:.4f}, Test RMSE: {test_rmse:.4f}")
        print(f"  Train R²: {train_r2:.4f}, Test R²: {test_r2:.4f}")
        print(f"  CV R²: {cv_scores.mean():.4f} (±{cv_scores.std():.4f})")
        print()
    
    return results

def plot_model_comparison(results):
    """Plot model comparison"""
    model_names = list(results.keys())
    test_r2_scores = [results[name]['test_r2'] for name in model_names]
    test_rmse_scores = [results[name]['test_rmse'] for name in model_names]
    
    fig, axes = plt.subplots(1, 2, figsize=(15, 6))
    
    # R² scores
    axes[0].barh(model_names, test_r2_scores)
    axes[0].set_xlabel('R² Score')
    axes[0].set_title('Model Comparison: R² Score')
    axes[0].grid(axis='x', alpha=0.3)
    
    # RMSE scores
    axes[1].barh(model_names, test_rmse_scores)
    axes[1].set_xlabel('RMSE')
    axes[1].set_title('Model Comparison: RMSE')
    axes[1].grid(axis='x', alpha=0.3)
    
    plt.tight_layout()
    plt.savefig('ml_model_comparison.png', dpi=300, bbox_inches='tight')
    plt.show()

def feature_importance_analysis(model, feature_names):
    """Analyze feature importance"""
    if hasattr(model, 'feature_importances_'):
        importances = model.feature_importances_
    elif hasattr(model, 'coef_'):
        importances = np.abs(model.coef_)
    else:
        print("Model does not support feature importance")
        return
    
    # Create DataFrame
    importance_df = pd.DataFrame({
        'Feature': feature_names,
        'Importance': importances
    }).sort_values('Importance', ascending=False)
    
    # Plot
    plt.figure(figsize=(10, 6))
    plt.barh(importance_df['Feature'], importance_df['Importance'])
    plt.xlabel('Importance')
    plt.title('Feature Importance')
    plt.tight_layout()
    plt.savefig('ml_feature_importance.png', dpi=300, bbox_inches='tight')
    plt.show()
    
    return importance_df

def hyperparameter_tuning(X_train, y_train, model_type='random_forest'):
    """Perform hyperparameter tuning"""
    print(f"\n=== HYPERPARAMETER TUNING: {model_type.upper()} ===\n")
    
    if model_type == 'random_forest':
        model = RandomForestRegressor(random_state=42)
        param_grid = {
            'n_estimators': [50, 100, 200],
            'max_depth': [10, 20, None],
            'min_samples_split': [2, 5, 10]
        }
    elif model_type == 'xgboost':
        model = xgb.XGBRegressor(random_state=42)
        param_grid = {
            'n_estimators': [50, 100, 200],
            'max_depth': [3, 5, 7],
            'learning_rate': [0.01, 0.1, 0.2]
        }
    else:
        print("Model type not supported for tuning")
        return None
    
    grid_search = GridSearchCV(
        model, param_grid, cv=5, scoring='r2', n_jobs=-1, verbose=1
    )
    grid_search.fit(X_train, y_train)
    
    print(f"Best parameters: {grid_search.best_params_}")
    print(f"Best CV score: {grid_search.best_score_:.4f}")
    
    return grid_search.best_estimator_

if __name__ == "__main__":
    # Load and prepare data
    df = load_data()
    df_clean = clean_data(df)
    
    X_train, X_test, y_train, y_test, scaler, feature_cols = prepare_ml_data(df_clean)
    
    print(f"Training set size: {X_train.shape}")
    print(f"Test set size: {X_test.shape}")
    print(f"Features: {feature_cols}\n")
    
    # Train and evaluate models
    results = train_and_evaluate_models(X_train, X_test, y_train, y_test)
    
    # Plot comparison
    plot_model_comparison(results)
    
    # Feature importance for best model
    best_model_name = max(results.keys(), key=lambda x: results[x]['test_r2'])
    best_model = results[best_model_name]['model']
    print(f"\nBest Model: {best_model_name}")
    importance_df = feature_importance_analysis(best_model, feature_cols)
    print("\nFeature Importance:")
    print(importance_df)
    
    # Hyperparameter tuning
    tuned_model = hyperparameter_tuning(X_train, y_train, 'random_forest')

